Partial corr optimisations 2

fredmanglis opened the pull request:
Commit cdb1d927f73339925d661dfa29b7a70522db5499 enables the use of Python's **multiprocessing** library/module to help with speeding up the computation of partial correlations.

The other 2 commits do some minor cleanup

@BonfaceKilz and/or @arunisaac Please review to merge this.

@zsloan This adds a new configuration variable, `MULTIPROCESSOR_PROCS` that limits the number of processes to spawn. I have initialised the number to 6, but you can change it to fit your deployment. Setting it to `None` will make the system spawn a number of processes that is one less than the total number of available processors.

fredmanglis committed:
Remove unnecessary computation

In Python3 when slicing,

     seq[:min(some_val, len(seq))] == seq[:some_val]

because Python3 will just return a copy of the entire sequence if `some_val`
happens to be larger/greater than the length of the sequence.

This commit removes the unnecessary call to `min()`

fredmanglis committed:
Use multiprocessing to speed up computation

This commit refactors the code to make it possible to use multiprocessing to
speed up the computation of the partial correlations.

The major refactor is to move the `__compute_trait_info__` function to the
top-level of the module, and provide to it all the other necessary context via
the new args.

fredmanglis committed:
Give sorting functions more descriptive names

BonfaceKilz commented on gn3/computations/partial_correlations.py:
```diff
@@ -296,57 +338,15 @@ def compute_partial(
-            __compute_trait_info__(target)
-            for target in zip(target_vals, target_names))
-        if result is not None)
+    with mp.Pool(MULTIPROCESSOR_PROCS or (mp.cpu_count() - 1)) as pool:
```
**nitpick: issue:** You could pass this as an arg with a default value; same goes for `partial_correlations_normal` where this function, `compute_partial` is only used.  Should someone ever need to re-use this, you wouldn't have made that decision for them, that is, defining, `MULTIPROCESSOR_PROCS`.  In this context, the final caller of this function should pick this value from the settings file and pass that in as an arg.

fredmanglis replied:
Well, that makes sense and I considered it - my thoughts though were more for the pragmatic (cough! lazy! cough!) approach, where I don't try to predict usage - especially for something this specific - and thus avoid painting myself into a corner.

BonfaceKilz commented on gn3/settings.py:
```diff
@@ -57,3 +57,5 @@ def parse_env_cors(default):
 
 ROUND_TO = 10
+
+MULTIPROCESSOR_PROCS = 6 # Number of processes to spawn
```
**nitpick:**  @zsloan and @fredmanglis we should make the default for this and make the value that works well on prod the default ;)

fredmanglis replied:
Yes indeed! @zsloan You can experiment and change the value, I guess...

BonfaceKilz reviewed (approved):
Just some minor nitpicks, but @fredmanglis this LGTM.  +1 for using the python 3.9's multiprocessing lib.  @Alexanderlacuna take note...

BonfaceKilz commented:
Muriithi Frederick Muriuki

[...]

>
> Well, that makes sense and I considered it - my thoughts though were more for the pragmatic (cough! lazy! cough!) approach, where I don't try to predict usage - especially for something this specific - and thus avoid painting myself into a corner.

I see.  Priority is having something that works.
And this is along that direction so +1 for that.
We should really start tracking our technical
debts(cough! lazy!  cough!) on gem-text ;)

-- 
Bonface M. K.
D4F09EB110177E03C28E2FE1F5BBAE1E0392253F (hkp://keys.gnupg.net)
Free Software Activist
Humble GNU Emacs User | Bearer of scheme-y parens
Curator: <https://upbookclub.com> | Twitter: @BonfaceKilz

Alexanderlacuna commented:
@fredmanglis @BonfaceKilz more like a general comment. While also attempting the same(multiprocessing) for correlation and also in this https://github.com/genenetwork/genenetwork3/pull/66/files I frequently
run into the same memory error in forked processes for huge datasets. Also, Pjotr pointed out that we already run gn2 in processes
so we are likely to run into nested processes.

fredmanglis commented:
@Alexanderlacuna @BonfaceKilz 

> ... I frequently run into the same memory error in forked processes for huge datasets. ...

Noted. Pjotr also [mentioned this on matrix](https://matrix.to/#/!GOCWkEcMnQruNBcCIN:matrix.org/$yjrlSDHp2CcHl-tyUIDG6Qex3lgVH5Rph9FYBOyRJP8?via=matrix.org).

I guess the only problem here is that using external processes involves communication between the two systems - meaning whatever gains I got from this will probably be swamped by the IO in the communication.

Back to the drawing board I guess. Any ideas are welcome.

fredmanglis committed:
Remove multiprocessing for stability

Web servers are long-running processes, and python is not very good at
cleaning up after itself especially in forked processes - this leads to memory
errors in the web-server after a while.

This commit removes the use of multiprocessing to avoid such failures.

fredmanglis committed:
Merge iterations to remove unnecessary computations

Do all the work in a single iteration to avoid unnecessary iterations that
hamper performance.

fredmanglis commented:
@Alexanderlacuna @BonfaceKilz I have removed the use of multiprocessing from the code, and found another tiny optimisation point.

If there are no objections, I am okay to have this merged.

BonfaceKilz commented:
Muriithi Frederick Muriuki
***@***.***> anaandika:

> @Alexanderlacuna @BonfaceKilz 
>
>> ... I frequently run into the same memory error in forked processes for huge datasets. ...
>
> Noted. Pjotr also [mentioned this on matrix](https://matrix.to/#/!GOCWkEcMnQruNBcCIN:matrix.org/$yjrlSDHp2CcHl-tyUIDG6Qex3lgVH5Rph9FYBOyRJP8?via=matrix.org).
>
> I guess the only problem here is that using external processes involves communication between the two systems - meaning whatever gains I got from this will probably be swamped by the IO in the communication.
>
> Back to the drawing board I guess. Any ideas are welcome.

I don't know if this is practical; but once this
is done and Rob is happy(at least for non-huge
datasets), you could do a re-write in something
like Rust and call it as an external process.
Similar to what we do with gemma and ChFi's
purescript genetics browser.  Just my 2¢.

-- 
Bonface M. K.
D4F09EB110177E03C28E2FE1F5BBAE1E0392253F (hkp://keys.gnupg.net)
Free Software Activist
Humble GNU Emacs User | Bearer of scheme-y parens
Curator: <https://upbookclub.com> | Twitter: @BonfaceKilz

robwwilliams commented:
If all is good for up to 5000 cases now then don’t worry too much if gets
slow. But we do have data sets for partial correlation that will be up to
10000 soon for our BIG children cohort.

Rust of maybe Julia.

On Tue, Feb 8, 2022 at 01:11 BonfaceKilz ***@***.***> wrote:

> Muriithi Frederick Muriuki
> ***@***.***> anaandika:
>
> > @Alexanderlacuna @BonfaceKilz
> >
> >> ... I frequently run into the same memory error in forked processes for
> huge datasets. ...
> >
> > Noted. Pjotr also [mentioned this on matrix](
> https://matrix.to/#/!GOCWkEcMnQruNBcCIN:matrix.org/$yjrlSDHp2CcHl-tyUIDG6Qex3lgVH5Rph9FYBOyRJP8?via=matrix.org).
>
> >
> > I guess the only problem here is that using external processes involves
> communication between the two systems - meaning whatever gains I got from
> this will probably be swamped by the IO in the communication.
> >
> > Back to the drawing board I guess. Any ideas are welcome.
>
> I don't know if this is practical; but once this
> is done and Rob is happy(at least for non-huge
> datasets), you could do a re-write in something
> like Rust and call it as an external process.
> Similar to what we do with gemma and ChFi's
> purescript genetics browser. Just my 2¢.
>
> --
> Bonface M. K.
> D4F09EB110177E03C28E2FE1F5BBAE1E0392253F (hkp://keys.gnupg.net)
> Free Software Activist
> Humble GNU Emacs User | Bearer of scheme-y parens
> Curator: <https://upbookclub.com> | Twitter: @BonfaceKilz
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/genenetwork/genenetwork3/pull/76#issuecomment-1032280324>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC55V7BUNHWNOGLWQSMSKA3U2C6ZFANCNFSM5NNXKRKA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
-- 
Rob

Robert W Williams PhD
Department of Genetics, Genomics and Informatics
71 S Manassas St, Memphis TN 38163
University of Tennessee Health Science Center, TSRB 405
 Cell: 901 604 4752
EMAIL: ***@***.***
Zoom: https://tennesseehipaa.zoom.us/j/9492269658
----------------
